### PR TITLE
[Preview] Item and Query Request Options: Change Dedicated Gateway MaxIntegratedCacheStaleness type in the header

### PR DESCRIPTION
**Pull request by soeom**
# Pull Request Template

## Description

To make MaxIntegratedCacheStaleness value language agnostic, we take TimeSpan as an input from the customer but send TotalMilliseconds in the header instead.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 1]
```diff
@@ -23,15 +23,16 @@ class DedicatedGatewayRequestOptions
         /// </summary> 
         /// <remarks> 
         /// For requests where the <see cref="ConsistencyLevel"/> is <see cref="ConsistencyLevel.Eventual"/>, responses from the integrated cache are guaranteed to be no staler than value indicated by this MaxIntegratedCacheStaleness. 
-        /// Value defaults to null. 
+        /// Value defaults to null. Cache Stalenss is supported in milliseconds granularity. Anything smaller than milliseconds will be ignored.
```
Comment (kushagraThapar) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:26]: Typo in `Stalenss` -> `Staleness`. 
Also, does default as null means Responses will not be retrieved from Cache ?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 2]
```diff
@@ -23,15 +23,16 @@ class DedicatedGatewayRequestOptions
...
         /// </remarks> 
         public TimeSpan? MaxIntegratedCacheStaleness { get; set; }
 
         internal static void PopulateMaxIntegratedCacheStalenessOption(DedicatedGatewayRequestOptions dedicatedGatewayRequestOptions, RequestMessage request)
         {
             if (dedicatedGatewayRequestOptions?.MaxIntegratedCacheStaleness != null)
             {
-                request.Headers.Set(HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness, dedicatedGatewayRequestOptions.MaxIntegratedCacheStaleness.Value.ToString("c", CultureInfo.InvariantCulture));
+                double cacheStalenessInMilliseconds = (double)dedicatedGatewayRequestOptions?.MaxIntegratedCacheStaleness.Value.TotalMilliseconds;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:34]: Please add a check for negative values.

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 1 above)
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:26]: I think the correct notation for the default value is `<value>Default value is null</value>`, like https://github.com/Azure/azure-cosmos-dotnet-v3/blob/765dcc58cb14b3291bf9162f31b93bddcc2b5a82/Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs#L130

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 2 above)
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:34]: Check for unsupported granularity, for example: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/c8935ac2f864fb829f5d941dda07c74aec86a677/Microsoft.Azure.Cosmos/src/Resource/Settings/ChangeFeedPolicy.cs#L57-L66

Comment (kushagraThapar) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:34]: @j82w - Can customers set negative values using TimeSpan ?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:34]: Yes, https://docs.microsoft.com/en-us/dotnet/api/system.timespan.negate?redirectedfrom=MSDN&view=net-5.0#System_TimeSpan_Negate

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/HandlerTests.cs [hunk 3]
```diff
@@ -216,7 +216,7 @@ public async Task QueryRequestOptionsDedicatedGatewayRequestOptions()
             {
                 TestHandler testHandler = new TestHandler((request, cancellationToken) =>
                 {
-                    Assert.AreEqual(maxStaleness.ToString(), request.Headers[HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness]);
+                    Assert.AreEqual(maxStaleness.TotalMilliseconds.ToString(), request.Headers[HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness]);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/HandlerTests.cs:219]:
```suggestion
                    Assert.AreEqual(maxStaleness.TotalMilliseconds.ToString(CultureInfo.InvariantCulture), request.Headers[HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness]);

```

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 4]
```diff
@@ -21,17 +21,25 @@ class DedicatedGatewayRequestOptions
...
+        /// Cache Staleness is supported in milliseconds granularity. Anything smaller than milliseconds will be ignored.
         /// </remarks> 
         public TimeSpan? MaxIntegratedCacheStaleness { get; set; }
 
         internal static void PopulateMaxIntegratedCacheStalenessOption(DedicatedGatewayRequestOptions dedicatedGatewayRequestOptions, RequestMessage request)
         {
             if (dedicatedGatewayRequestOptions?.MaxIntegratedCacheStaleness != null)
             {
-                request.Headers.Set(HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness, dedicatedGatewayRequestOptions.MaxIntegratedCacheStaleness.Value.ToString("c", CultureInfo.InvariantCulture));
+                double cacheStalenessInMilliseconds = (double)dedicatedGatewayRequestOptions?.MaxIntegratedCacheStaleness.Value.TotalMilliseconds;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:35]: Already checked for null. No need for it.
```suggestion
                double cacheStalenessInMilliseconds = (double)dedicatedGatewayRequestOptions.MaxIntegratedCacheStaleness.Value.TotalMilliseconds; 
```

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 5]
```diff
@@ -21,17 +21,25 @@ class DedicatedGatewayRequestOptions
         /// <summary> 
         /// Gets or sets the staleness value associated with the request in the Azure CosmosDB service. 
         /// </summary> 
+        /// <value>Default value is null.</value> 
         /// <remarks> 
         /// For requests where the <see cref="ConsistencyLevel"/> is <see cref="ConsistencyLevel.Eventual"/>, responses from the integrated cache are guaranteed to be no staler than value indicated by this MaxIntegratedCacheStaleness. 
-        /// Value defaults to null. 
+        /// Cache Staleness is supported in milliseconds granularity. Anything smaller than milliseconds will be ignored.
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:27]: nit: I don't think you can create a TimeSpan with a lower granularity than Milliseconds?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:27]: You can do it with ticks which are less than a Milliseconds

**Review by j82w — APPROVED** [review]

**Pull request merged**